### PR TITLE
Disallow non octal decimal escape before use strict

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Babel does not throw for `() => { "\8";"\9";"use strict"; }` ([REPL](https://babel.dev/repl#?browsers=chrome%208&build=&builtIns=false&spec=false&loose=false&code_lz=BQSgBAvAfGDeYCIA6AOBBuZBODCCuAzgKZgEAuATgJYDGZGYAvkA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=false&presets=env%2Creact%2Cstage-2%2Cflow&prettier=false&targets=&version=7.12.3&externalPlugins=))
| Tests Added + Pass?      | Yes
| License                  | MIT

This PR also refactors how we track numeric escape (though named as `octalPositions`). Currently `this.state.octalPositions` is reset whenever `this.nextToken` is called, this is unnecessary and can be avoided by only reset when a) directives are allowed in blocks and b) we have not yet seen a non-directive parsed. I renamed `octalPositions` to `strictErrors` since we need to track two errors in this PR

- StrictOctalLiteral
- StrictNumericEscape

However I think we should throw StrictNumericEscape for both `\8` and `\01`, since the latter is not an octal literal but an octal escape sequence, which falls into numeric escape. I have update the error message in this PR.

This PR also avoids reparsing the numeric/string literal after strict directive in `setState`, since the octal error is now tracked in `strictErrors`.

Last this PR extracts legacy octal literal test cases from string escape sequence tests.

@kaicataldo Feedbacks are welcome since this PR is a follow-up to & inspired from #11188.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12366"></a>

### CODE REVIEW

File: packages/babel-parser/src/tokenizer/index.js [hunk 1]
```diff
@@ -1393,13 +1404,11 @@ export default class Tokenizer extends ParserErrors {
...
-            } else if (this.state.strict) {
-              this.raise(codePos, Errors.StrictNumericEscape);
             } else {
-              // This property is used to throw an error for
-              // an octal literal in a directive that occurs prior
-              // to a "use strict" directive.
-              this.state.octalPositions.push(codePos);
+              this.recordNumericEscapeErrors(
+                codePos,
+                Errors.StrictOctalLiteral,
```
Comment (JLHwung) [inline packages/babel-parser/src/tokenizer/index.js:1410]: We should have thrown StrictNumericEscape since we are parsing an escaped sequence. Will addressed it in another PR.

File: scripts/parser-tests/test262/allowlist.txt [hunk 2]
```diff
@@ -1 +1,2 @@
-
+language/literals/string/legacy-non-octal-escape-sequence-8-strict-explicit-pragma.js(default)
+language/literals/string/legacy-non-octal-escape-sequence-9-strict-explicit-pragma.js(default)
```
Comment (JLHwung) [inline scripts/parser-tests/test262/allowlist.txt]: This is bug is caught by these two failing test262 tests.

File: packages/babel-parser/test/fixtures/core/escape-string/invalid-decimal-escape-strict-directive-function/input.js [hunk 3]
```diff
@@ -20,13 +20,3 @@ function d() {
...
-function c() {
-  "use strict";
-  05;
-}
-
-function d() {
-  "use strict";
-  04;
-  05;
-}
```
Comment (JLHwung) [inline packages/babel-parser/test/fixtures/core/escape-string/invalid-decimal-escape-strict-directive-function/input.js:32]: They are moved to https://github.com/babel/babel/pull/12366/files#diff-cd40ee2fb191cbf29a8bbfc49ae7471de322ff395f79acf129d3f9cd3832a688

So this test is focused on numeric escape sequence only.

File: packages/babel-parser/src/tokenizer/state.js [hunk 4]
```diff
@@ -137,10 +137,15 @@ export default class State {
...
-  // to a "use strict" directive.
-  octalPositions: number[] = [];
+  // This property is used to track the following errors
+  // - StrictNumericEscape
+  // - StrictOctalLiteral
+  //
+  // in a literal that occurs prior to/immediately after a "use strict" directive.
+
+  // todo(JLHwung): set strictErrors to null and avoid recording string errors
+  // after a non-directive is parsed
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/tokenizer/state.js:147]: Do you want to do this here or in a separate PR?

Comment (JLHwung) [reply packages/babel-parser/src/tokenizer/state.js:147]: I prefer a separate PR since alternatively, todo is pronounced as never-do. 😛 
Legacy octal or numeric escapes are discouraged, I feel like it is okay to leave it as-is.

**Review by nicolo-ribaudo — APPROVED** [review]

**Review by existentialism — APPROVED** [review]
💯
